### PR TITLE
Components: Add hooks based VisuallyHidden

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Description
Create a new `VisuallyHidden` based on the existing `VisuallyHidden`'s styles but use `createComponent` and a new `useVisuallyHidden` hook.

## How has this been tested?
Snapshot tests updated and tested in storybook.

### Focused

Captura de Tela 2021-02-09 às 08 53 10

### Normal 

Captura de Tela 2021-02-09 às 08 53 25

## Types of changes
New feature.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by gziolo** [comment]
I couldn't find CSS styles for the focused state of the `VisuallyHidden` component. Are they integrated into the new implementation?

**Comment by saramarcondes** [comment]
@gziolo Thanks for pointing that out, they're indeed missing, it'll be trivial to add them using the `css` prop, however. Fix incoming 👍🏼

**Comment by gziolo** [comment]
I repeat myself, but those snapshot tests aren't very helpful. Those diffs are so gigantic that it's impossible to validate the change if you don't know the context. Therefore they don't bring too much value to the review process. We should still use them but find a better way to distill what's saved to only the bits that are important for the individual test.

**Comment by saramarcondes** [comment]
@ItsJonQ Do you have any ideas what to do about the snapshot tests? I'll do a little research today to see how others have solved this issue but so far all I've seen is advise to use the emotion serializer (which we do).

**Comment by gziolo** [comment]
Diffing between the most basic version would be a great start.

I found this article where it's recommended: https://kentcdodds.com/blog/effective-snapshot-testing

**Comment by ItsJonQ** [comment]
> Do you have any ideas what to do about the snapshot tests? 

@saramarcondes Not for the time being :(. Other than to skim them to ensure they look correct, and update if necessary (what we're doing now). This is of course more difficult if lower level changes affect many other snapshots.

As @saramarcondes had suggested (and Kent had mentioned in his blog post), maybe `snapshot-diff` can help.

---

Taking a step back, replacing the current `VisuallyHidden` with the G2 setup will mark the first substantial component we'd be refactoring. FontSizePicker happened, but it's footprint is very minimal within Gutenberg, and it is controlled with the adapters.

This `VisuallyHidden` update is different as it updates the current component entirely without adapters.

Having chatted with @saramarcondes , I think this would be a good candidate to increase G2 Component exposure while being mindful of risk. In theory, `VisuallyHidden` has very little risk as it doesn't render anything on the UI - that's it's job after all. The chances of this breaking other implements is very minimal.

For that, I'm feeling okay moving forward with this.

@gziolo Love to hear your thoughts :)

**Comment by saramarcondes** [comment]
I think I’ve got something close to working with snapshot diff:

```
exports[`props should render align 1`] = `
Snapshot Diff:
- First value
+ Second value

@@ -35,11 +35,11 @@
    "_values": Object {
      "--wp-g2-flex-gap": "calc(var(--wp-g2-grid-base) * 2)",
      "--wp-g2-flex-item-margin-bottom": "0",
      "--wp-g2-flex-item-margin-left": "0",
      "--wp-g2-flex-item-margin-right": "var(--wp-g2-flex-gap)",
-     "align-items": "flex-start",
+     "align-items": "center",
      "box-sizing": "border-box",
      "display": "flex",
      "flex-direction": "row",
      "font-family": "var(--wp-g2-font-family)",
      "font-weight": "var(--wp-g2-font-weight)",
`;
```

Only problem is snapshot diff does not play nicely with emotion (like at all) so I think we’ll need to write a custom matcher for it.

The code for the test above is this:

```
const { container } = render(
	<Flex align="flex-start">
		<FlexItem />
		<FlexBlock />
	</Flex>
);
expect(
	window.getComputedStyle( container.firstChild )
).toMatchDiffSnapshot(
	window.getComputedStyle( base.container.firstChild )
);
```

Where `base` is the same as `container` but without `align` set.

The custom matcher could be named something like `toMatchStyleDiffSnapshot` and would pass the expected and received values through `getComputedStyle` (or rather, whatever `toHaveStyle` uses because that works with emotion).

In any case, progress is being made here, just wanted to let y'all know. I should have a PR open soon with a proposal for this, either today or tomorrow.

**Comment by ItsJonQ** [comment]
> I should have a PR open soon with a proposal for this, either today or tomorrow.
@saramarcondes Awesome! Excited to see what you have in mind 😍

**Comment by saramarcondes** [comment]
Here y'all go: https://github.com/WordPress/gutenberg/pull/28897

**Comment by gziolo** [comment]
I would prefer we don't remove old code until there is an official proposal that is accepted. It should be a small change here.

**Comment by sarayourfriend** [comment]
Sounds good, we'll just move this into the `ui` folder and restore the original.

File: packages/components/src/ui/visually-hidden/use-visually-hidden.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import * as styles from './visually-hidden-styles';
+
+/**
+ *
+ * @param {import('@wp-g2/create-styles').ViewOwnProps<{}, 'div'>} props
+ */
+export function useVisuallyHidden( { className, ...props } ) {
+	// circumvent the context system and write the classnames ourselves
+	const classes = cx(
+		'components-visually-hidden wp-components-visually-hidden',
```
Comment (gziolo) [inline packages/components/src/ui/visually-hidden/use-visually-hidden.js:18]: Out of curiosity, why does it have two similar class names?

The old implementation uses only `components-visually-hidden`.

**Comment by gziolo** [comment]
Some things I observe here and in general. Those are nitpicking to consider but they might help with further work:
- other components expose both the component and its hook, I wouldn't mind if you would keep it the same for everything if only possible
- for other components `index.js` only re-exports public API, the hook is in the `use-component-name.js` and the component is in `component-name.js`, should we standardize it and simplify the naming of the files, how do you feel about:
  - `index.js`
  - `component.js`
  - `hook.js`
  - I wouldn't mind if we wouldn't repeat the name of the component in paths like `component-name/component-name-*.js`
- can we have at least one unit test for every component? in this case, is it even possible to validate if the component isn't visible with all the testing tools? a snapshot could be fine otherwise 😄

File: packages/components/src/ui/visually-hidden/use-visually-hidden.js (hunk 1 above)
Comment (sarayourfriend) [reply packages/components/src/ui/visually-hidden/use-visually-hidden.js:18]: G2 introduces the `wp-` prefixed classname. Typically they'd be added by `useContextSystem` here: https://github.com/ItsJonQ/g2/blob/5feb4f2822efcd67afbf3f52849c78b8cbfd0f82/packages/context/src/use-context-system.js#L95

For consistency's sake I added both to the hooks version.

**Comment by sarayourfriend** [comment]
> Should we connect the new component using withNext liker here:

I don't think this is necessary considering there is no need for a props adapter, we can just swap out the references directly, right?

Or should we do it so that we can swap them using the context system globally?

**Comment by ItsJonQ** [comment]
> ...should we standardize it and simplify the naming of the files

I'm happy with anything.

For context, I'm personally use to naming conventions I have in the [`@itsjonq/g2` repo](https://github.com/ItsJonQ/g2/tree/main/packages/components/src/Flex) (which I mirrored from [other](https://github.com/Shopify/polaris-react/tree/main/src/components/Button) [projects](https://github.com/primer/components/tree/main/src) in this [component library](https://github.com/adobe/react-spectrum/tree/main/packages/%40react-spectrum/buttongroup/src) space). It doesn't matter though 😊 .

The only thing the proposed naming system doesn't address is sub-components 🤔 

(Let's pretend there's a `list` component)

`src/components/ui/list/`
* `index.js`
* `component.js`
* `hook.js`
* `utils.js` (maybe?)
* `list-item.js` (or would it be `item.js`)

(I think `list-item.js` would be easiest)

> Should we connect the new component using withNext liker here:

@gziolo Since we're adding this (new) `VisuallyHidden` under the `/ui` directory, rather than replacing the existing one, I think it makes sense to wrap the existing one with the `withNext()` HOC so we can migrate it

**Comment by gziolo** [comment]
It's a tricky part to name files. We were using mostly `index.js` in other places and then the name of the component would match the folder name. If there are other components then usually you will see the unique part of the component name as the name of the file. I'm mostly thinking about consistency so we you can quickly get the pattern when extending functionality.

**Comment by sarayourfriend** [comment]
@gziolo @ItsJonQ Does this look good to go?

**Comment by gziolo** [comment]
Screen Shot 2021-02-12 at 17 56 48

Any idea where this error comes from? Would rebase help? Otherwise, it's ✅ from me.

**Review by gziolo — APPROVED** [review]

**Pull request merged**